pulp_container/app/registry_api.py

```diff
@@ -475,6 +479,161 @@ def get_queryset(self, *args, **kwargs):
         return (public_repositories | accessible_repositories).distinct()
 
 
+class FlatpakIndexDynamicView(APIView):
+    """
+    Handles requests to the /index/dynamic endpoint
+    """
+
+    def get(self, request):
+        unsatisfiable = False
+        req_repository = None
+        req_tags = set()
+        req_os = None
+        req_architecture = None
+        req_label_exists = {"org.flatpak.ref"}
```

Comment:
should the second label be added here too? https://github.com/owtaylor/flatpak-indexer/blob/main/flatpak_indexer/indexer.py#L85

Reply:
With "second label" you refer to `k.startswith('org.freedesktop.')`, i.e., initialize `req_label_exists` to `{"org.flatpak.ref", "org.freedesktop.ref"}`?  That wouldn't work, as it would require that any images to be returned would need to have both some `org.flatpak.ref` and some `org.freedesktop.ref` label.  But e.g. <https://registry.fedoraroject.org/index/static> only reports `org.flatpak.ref` labels and no `org.freedesktop.ref` labels at all.  So if we import images from that flatpak registry into pulp, none of the imported images would have an `org.freedesktop.ref` label, and this endpoint would in turn not report any images to clients.
And I think the flatpak-indexer code you reference is about including both `org.flatpak...` and `org.freedesktop...` (like `org.freedesktop.appstream.appdata`) Koji labels, not about support for some hypothetical `org.freedesktop.ref` label.
(The code above, `req_label_exists = {"org.flatpak.ref"}`, is about presetting the set of requested `label:...:exists=1` query parameters with `label:org.flatpak.ref:exists=1`, in case a client would fail to include that in its query.  I thought that <https://github.com/owtaylor/flagstate/blob/master/docs/protocol.md> mentions that that should always be included in queries, but it looks like it doesn't actually mention that, and that I only learned from private communication with @owtaylor that (a) clients should always include `label:org.flatpak.ref:exists=1` in their queries and (2) servers should always assume that `label:org.flatpak.ref:exists=1` is implicitly assumed for every query.  Maybe we should add something about that to <https://github.com/owtaylor/flagstate/blob/master/docs/protocol.md>.)

Reply:
> should the second label be added here too? https://github.com/owtaylor/flatpak-indexer/blob/main/flatpak_indexer/indexer.py#L85

This code is specifically about converting the "new" type of Flatpaks with metadata in labels to be compatible with older clients (older clients can be ignored at this point.) In terms of creating labels, there is no special handling for any labels - all should be included.

> (The code above, req_label_exists = {"org.flatpak.ref"}, is about presetting the set of requested label:...:exists=1 query parameters with label:org.flatpak.ref:exists=1, in case a client would fail to include that in its query. [...]  I only learned from private communication with @owtaylor that (a) clients should always include label:org.flatpak.ref:exists=1 in their queries and (2) servers should always assume that label:org.flatpak.ref:exists=1 is implicitly assumed for every query. Maybe we should add something about that to https://github.com/owtaylor/flagstate/blob/master/docs/protocol.md.)

I think what I meant was more along the lines of (1) since the only client that talks this protocol is the Flatpak client, it *will* always include `label:org.flatpak.ref:exists=1`, and (2) since the only clients that talk this protocol is the Flatpak client, servers *can* assume that org.flatpak.ref:exists=1.

(For Fedora and RHEL, we pre-generate indexes for this case, and 404 if we get a query without the restriction.)

But unless the 2) actually simplifies things, you definitely don't need to make that assumption.